examples/pytorch/lda/lda_model.py

```diff
@@ -98,133 +79,207 @@ class LatentDirichletAllocation:
-        doc_data = _update_all(
-            G, 'doc', self.prior['doc'], self.step_size['doc'],
-            return_obj=True)
+        if doc_data is None:
+            doc_data = self._e_step(G)
+        word_data = {'z': self.word_z, 'weight': self._word_weight}
 
         # compute E[log p(docs | theta, beta)]
-        edge_elbo = (
-            doc_data['edge_elbo'].sum() / doc_data['z'].sum()
-        ).cpu().numpy()
+        edge_elbo = self._edge_elbo(G, doc_data).cpu().numpy()
         if self.verbose:
             print(f'neg_elbo phi: {-edge_elbo:.3f}', end=' ')
 
         # compute E[log p(theta | alpha) - log q(theta | gamma)]
-        doc_elbo = (
-            (-doc_data['z'] * doc_data['weight'].log()).sum(axis=1)
-            -_lbeta(self.prior['doc'] + doc_data['z'] * 0, axis=1)
-            +_lbeta(self.prior['doc'] + doc_data['z'], axis=1)
-        )
-        doc_elbo = (doc_elbo.sum() / doc_data['z'].sum()).cpu().numpy()
+        doc_elbo = self._node_elbo(doc_data, 'doc').cpu().numpy()
         if self.verbose:
             print(f'theta: {-doc_elbo:.3f}', end=' ')
 
-        # compute E[log p(beta | eta) - log q (beta | lambda)]
-        word_elbo = (
-            (-word_data['z'] * word_data['weight'].log()).sum(axis=0)
-            -_lbeta(self.prior['word'] + word_data['z'] * 0, axis=0)
-            +_lbeta(self.prior['word'] + word_data['z'], axis=0)
-        )
-        word_elbo = (word_elbo.sum() / word_data['z'].sum()).cpu().numpy()
+        # compute E[log p(beta | eta) - log q(beta | lambda)]
+        # The denominator z.sum() for extrapolation perplexity is undefined.
+        # We use the train set, whereas sklearn uses the test set.
+        word_elbo = self._node_elbo(word_data, 'word').cpu().numpy()
         if self.verbose:
             print(f'beta: {-word_elbo:.3f}')
 
-        return np.exp(-edge_elbo - doc_elbo - word_elbo)
+        ppl = np.exp(-edge_elbo - doc_elbo - word_elbo)
+        if G.num_edges()>0 and np.isnan(ppl):
+            warnings.warn("numerical issue in perplexity")
+        return ppl
 
 
 if __name__ == '__main__':
-    print('Testing LatentDirichletAllocation via task_example_test.sh ...')
-    tf_uv = np.array(np.nonzero(np.random.rand(20,10)<0.5)).T
-    G = dgl.heterograph({('doc','topic','word'): tf_uv.tolist()})
+    print('Testing LatentDirichletAllocation ...')
+    G = dgl.heterograph({('doc', '', 'word'): [(0, 0), (0, 3)]})
     model = LatentDirichletAllocation(G, 5, verbose=False)
     model.fit(G)
     model.transform(G)
+    model.sample(model.transform(G), 2)
     model.perplexity(G)
+
+    sampler = dgl.dataloading.MultiLayerFullNeighborSampler(1)
+    dataloader = dgl.dataloading.NodeDataLoader(
+        G.reverse(), {'doc': np.arange(G.num_nodes('doc'))}, sampler,
+        batch_size=1024, shuffle=True, drop_last=False)
+    for input_nodes, _, (block,) in dataloader:
+        B = dgl.DGLHeteroGraph(
+            block._graph, ['_', 'word', 'doc', '_'], block.etypes
```

Comment:
Looks like this is constructing a normal graph from a block.  Does `dgl.block_to_graph` work for you?

Reply:
I wanted to only get "word_src->doc_dst" and then rename the nodes back to "word" and "doc". What is the best way to do that?

Reply:
OK I just realized that there's not an easy way to rename a graph's node and edge types.  Though technically you could call `block_to_graph` and then reconstruct another graph with the same edges.  I'll leave it like this for now.

Reply:
"call block_to_graph and then reconstruct another graph with the same edges" <- how?